homeassistant/components/lamarzocco/config_flow.py

```diff
@@ -0,0 +1,156 @@
+                        self.reauth_entry, data=user_input
+                    )
+                    await self.hass.config_entries.async_reload(
+                        self.reauth_entry.entry_id
+                    )
+                    return self.async_abort(reason="reauth_successful")
+
+            if not errors:
+                self._config = data
+                return await self.async_step_machine_selection()
+
+        return self.async_show_form(
+            step_id="user",
+            data_schema=vol.Schema(
+                {
+                    vol.Required(CONF_USERNAME): str,
+                    vol.Required(CONF_PASSWORD): str,
+                }
+            ),
+            errors=errors,
+        )
+
+    async def async_step_machine_selection(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Let user select machine to connect to."""
+        errors: dict[str, str] = {}
+        if user_input:
+            serial_number = user_input[CONF_MACHINE]
+            await self.async_set_unique_id(serial_number)
+            self._abort_if_unique_id_configured()
+
+            # validate local connection if host is provided
+            if user_input.get(CONF_HOST):
+                lm = LaMarzoccoClient()
+                if not await lm.check_local_connection(
+                    credentials=self._config,
+                    host=user_input[CONF_HOST],
+                    serial=serial_number,
+                ):
+                    errors[CONF_HOST] = "cannot_connect"
+
+            if not errors:
+                return self.async_create_entry(
+                    title=serial_number,
+                    data=self._config | user_input,
+                )
+
+        machine_options = [
+            SelectOptionDict(
+                value=serial_number,
+                label=f"{model_name} ({serial_number})",
+            )
+            for serial_number, model_name in self._machines
+        ]
+
+        machine_selection_schema = vol.Schema(
+            {
+                vol.Required(
+                    CONF_MACHINE, default=machine_options[0]["value"]
+                ): SelectSelector(
+                    SelectSelectorConfig(
+                        options=machine_options,
+                        mode=SelectSelectorMode.DROPDOWN,
+                    )
+                ),
+                vol.Optional(CONF_HOST): cv.string,
+            }
+        )
+
+        return self.async_show_form(
+            step_id="machine_selection",
+            data_schema=machine_selection_schema,
+            errors=errors,
+        )
+
+    async def async_step_reauth(self, entry_data: Mapping[str, Any]) -> FlowResult:
+        """Perform reauth upon an API authentication error."""
+        self.reauth_entry = self.hass.config_entries.async_get_entry(
+            self.context["entry_id"]
+        )
+        return self.async_show_form(
```

Comment:
There must be a confirmation by the user before continuing the reauth flow. This step is called by default with the existing config entry data as input. Therefore we need a separate step that handles the user input.

Most existing integration has a separate confirm step for the reauth or calls the user step to handle the user input. Please search for examples among existing integrations, if the docs example isn't enough.

https://developers.home-assistant.io/docs/config_entries_config_flow_handler#reauthentication